drivers/adc/adc_tla2021.c

```diff
@@ -0,0 +1,330 @@
+
+	const struct tla2021_config *config = dev->config;
+	uint8_t value_[3] = {reg};
+
+	sys_put_be16(value, &value_[1]);
+
+	ret = i2c_write_dt(&config->bus, value_, sizeof(value_));
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline int read_data(const struct device *dev, tla2021_reg_data_t *value)
+{
+	return read_(dev, REG_DATA, value);
+}
+
+static inline int read_config(const struct device *dev, tla2021_reg_config_t *value)
+{
+	return read_(dev, REG_CONFIG, value);
+}
+
+static inline int write_config(const struct device *dev, tla2021_reg_config_t value)
+{
+	return write_(dev, REG_CONFIG, value);
+}
+
+static int tla2021_channel_setup(const struct device *dev, const struct adc_channel_cfg *cfg)
+{
+	if (cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid gain");
+		return -EINVAL;
+	}
+
+	if (cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Invalid reference");
+		return -EINVAL;
+	}
+
+	if (cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Invalid acquisition time");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int tla2021_start_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	if (seq->channels != BIT(0)) {
+		LOG_ERR("Selected channel(s) not supported: %x", seq->channels);
+		return -EINVAL;
+	}
+
+	if (seq->resolution != 12) {
+		LOG_ERR("Selected resolution not supported: %d", seq->resolution);
+		return -EINVAL;
+	}
+
+	if (seq->oversampling) {
+		LOG_ERR("Oversampling is not supported");
+		return -EINVAL;
+	}
+
+	if (seq->calibrate) {
+		LOG_ERR("Calibration is not supported");
+		return -EINVAL;
+	}
+
+	if (!seq->buffer) {
+		LOG_ERR("Buffer invalid");
+		return -EINVAL;
+	}
+
+	const size_t num_extra_samples = seq->options ? seq->options->extra_samplings : 0;
+	const size_t num_samples = (1 + num_extra_samples) * POPCOUNT(seq->channels);
+
+	if (seq->buffer_size < (num_samples * sizeof(tla2021_reg_data_t))) {
+		LOG_ERR("buffer size too small");
+		return -EINVAL;
+	}
+
+	struct tla2021_data *data = dev->data;
+
+	data->buffer = seq->buffer;
+
+	adc_context_start_read(&data->ctx, seq);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int tla2021_read_async(const struct device *dev, const struct adc_sequence *seq,
+			      struct k_poll_signal *async)
+{
+	int ret;
+
+	struct tla2021_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	ret = tla2021_start_read(dev, seq);
+	adc_context_release(&data->ctx, ret);
+
+	return ret;
+}
+
+static int tla2021_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	return tla2021_read_async(dev, seq, NULL);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	int ret;
+
+	struct tla2021_data *data = CONTAINER_OF(ctx, struct tla2021_data, ctx);
+	const struct device *dev = data->dev;
+
+	tla2021_reg_config_t reg = data->reg_config;
+
+	/*
+	 * Start single-shot conversion
+	 */
+	WRITE_BIT(reg, REG_CONFIG_MODE_pos, 1);
+	WRITE_BIT(reg, REG_CONFIG_OS_pos, 1);
```

Comment:
I think `WRITE_BIT` is a bit uncommon if there's no condition, how about just `reg |= BIT(REG_CONFIG_MODE_pos)`?

Reply:
What do you mean? `WRITE_BIT` comes from `util_macro.h`...

Reply:
I mean I find it odd to use `WRITE_BIT` with a fixed argument, like I normally think of this macro used to replace patterns like:

```
if (set) {
  val |= BIT(bit);
} else {
  val &= ~BIT(bit);
}
```

but you have no condition so why not just use `reg |= BIT(REG_CONFIG_MODE_pos) | BIT(REG_CONFIG_OS_pos)` in this case?

More of a preference, it just looks unusual to me.

Reply:
I see you point. But since it's static it doesn't add any complexity after compilation also manual bit operations would ad much more noise to the code.

Reply:
Cool with that.